File: 2021_lua/day01.lua
-- Advent Of Code 2021
-- Copyright (C) 2024  Lukas Heindl
-- 
-- This program is free software: you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation, either version 3 of the License, or
-- (at your option) any later version.
-- 
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
-- 
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

local _M = {}

function _M.part1(file)
	local ret = 0
	local n1,n2 = file:read("n"),file:read("n")
	while n2 do
		if n1 < n2 then ret = ret+1 end
		n1 = n2
		n2 = file:read("n")
	end
	return ret
end

function _M.part2(file)
	local ret = 0
	local lastSum = nil
	local sums = {0, 0, 0}
	local n = file:read("n")
	sums[1] = n
	n = file:read("n")
	sums[2] = n
	sums[1] = sums[1] + n
	n = file:read("n")
	while n do
		sums[1] = sums[1] + n
		sums[2] = sums[2] + n
		sums[3] = sums[3] + n

		-- sum[1] is finished
		if lastSum and lastSum < sums[1] then ret = ret+1 end
		lastSum = sums[1]
		table.remove(sums, 1)
		sums[3] = 0
		n = file:read("n")
	end
	return ret
end

return _M

File: 2022_Cpp/day01.cpp
/*
Advent Of Code 2020
Copyright (C) 2024  Lukas Heindl

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <algorithm>
#include <fstream>
#include <iostream>
#include <iterator>
#include <numeric>
#include <vector>

#include "aocutils.h"

struct Elf
{
	std::vector<int> bag;
	int              calories;

	operator int() { return calories; }

	friend std::istream &operator>>(std::istream &is, Elf &i)
	{
		i = std::move(Elf{});
		// std::cout << &i << std::endl;
		// std::cout << i.bag.size() << std::endl;
		std::string l;
		std::getline(is, l);
		while(l != "")
		{
			// std::cout << "'" << l << "'" << std::endl;
			i.bag.push_back(std::stoi(l));
			std::getline(is, l);
		}
		i.calories = std::reduce(i.bag.begin(), i.bag.end());
		return is;
	}
};

int main()
{
	// std::ifstream ifs{"../problems/day01.dat.testing"};
	std::ifstream ifs{"../problems/day01.dat"};
	if(ifs.fail()) { throw std::runtime_error("File couldn't be opened!"); }
	std::vector<Elf> elves = aocutils::vectorize_ifs<Elf>(ifs);

	std::sort(elves.begin(), elves.end(),
			  [](const auto &a, const auto &b)
			  { return a.calories < b.calories; });
	auto part2 = std::accumulate(elves.end() - 3, elves.end(), 0);
	auto part1 = (elves.end() - 1)->calories;
	// old stuff before the need of sorting the vector
	// auto x = std::ranges::max_element(elves.begin(), elves.end(), [](const
	// auto& a, const auto& b) {return a.calories < b.calories;});

	std::cout << "Day 01:" << '\n'
			  << "  Part 1: " << part1 << std::endl
			  << "  Part 2: " << part2 << std::endl;
}
